Remove config creator

KSGulin opened the pull request:
This PR removes the config creator from sparsify and replaces the functionality with calls to the Neural Magic API.

Note that in this PR the calls to the API are directed to local host. This address will be properly updated before merging sparsify.alpha into main, once the API goes live. 

**Testing**
covered by automatic pytest testing and integration tests

Ran a manual test by running ./bin/run-local within this repo, followed by sparsify.auto --task object_detection --data coco128.yaml in sparsify

KSGulin committed:
Initial commit, functional but not fully cleaned up or tested

KSGulin committed:
Initial commit, functional but not fully cleaned up or tested

corey-nm commented on src/sparsify/auto/api/main.py:
```diff
@@ -12,48 +12,65 @@
+            },
+        )
+
+        if response.json().get("tuning_complete"):
+            break
+
+        configs.append(SparsificationTrainingConfig(**response.json()))
```
Yeah as mentioned in other PR, I think doing something like:
```suggestion
        config_dict = response.json()
        if config_dict is None:
            break
        config = SparsificationTrainingConfig(**config_dict)
```
Could be more explicit. Unless there's something else that we may want to include in the completion dict?

KSGulin replied:
We want to be explicit that the return is a signal that early termination condition was met. None leaves too much room for ambiguity

corey-nm commented on src/sparsify/auto/api/main.py:
```diff
@@ -12,48 +12,65 @@
 
         # Execute integration run and build output object
-        runner_outputs = runner.run()
+        runner_outputs.append(runner.run())
```
I think moving to list of tuples could help this out. You could do:
```python
history = []
config = ... # initial request
while ...:
    runner = ...
    metrics = runner.run()
    history.append((config, metrics))
    response = requests.post(... json={'history': history})
```

KSGulin replied:
Good call, updated 🙂

corey-nm commented on src/sparsify/auto/api/main.py:
```diff
@@ -12,48 +12,65 @@
+        time.time() - training_start_time > max_train_seconds
+    ):
+        # request initial training config
+        if len(configs) == 0:
```
can you move this out of the loop to just always run?

corey-nm reviewed (changes requested):
Ack sorry meant to just comment instead approve

bfineran commented on src/sparsify/auto/api/main.py:
```diff
@@ -12,48 +12,65 @@
+        if len(configs) == 0:
+            # TODO: add proper server error handling. Waiting for server to have
+            # graceful failure protocol
+            response = requests.post(
+                f"{NEURAL_MAGIC_API_ADDRESS}/v1/sparsify/auto/training-config",
+                json=api_args.dict(),
```
let's move this into a helper function - prefer to separate API query handling from the main logic loop

KSGulin replied:
Good call, done

bfineran commented on src/sparsify/auto/api/main.py:
```diff
@@ -12,48 +12,65 @@
+        runner_outputs.append(runner.run())
+
+        # request a config with a new set of hyperparameters
+        response = requests.post(
+            f"{NEURAL_MAGIC_API_ADDRESS}/v1/sparsify/auto/training-config/tune",
+            json={
+                "configs": [config.dict() for config in configs],
+                "metrics": [output.metrics.dict() for output in runner_outputs],
+            },
+        )
```
ditto

bfineran commented on src/sparsify/auto/api/main.py:
```diff
@@ -12,48 +12,65 @@
+            },
+        )
+
+        if response.json().get("tuning_complete"):
```
what is the expected response model here? would it make sense to break into something more structured, or over-engineering for now?

KSGulin replied:
Over-engineering I think. We'd introduce an ambiguity of dict -> pydantic model assignment which could definitely be engineered around, but it doesn't more closer to the ideal than this also less than ideal approach

bfineran commented on tests/sparsify/auto/autosparse/test_registered_tasks.py:
```diff
@@ -11,19 +11,29 @@
+    )
+    assert response.status_code
 
+    assert [char in response.text for char in ["]", "["]]
```
took a second to read the list, not sure if `char in "[]"` would be any easier. nbd

bfineran commented on tests/sparsify/auto/autosparse/test_registered_tasks.py:
```diff
@@ -11,19 +11,29 @@
+    assert response.status_code
 
+    assert [char in response.text for char in ["]", "["]]
+    config_tasks = [TASK_REGISTRY[task] for task in ast.literal_eval(response.text)]
```
why do we need ast here?

bfineran reviewed (commented):
LGTM overall, minor comments

KSGulin committed:
Consolidate runs into history

KSGulin committed:
Consolidate runs into history

KSGulin committed:
minor test nit

KSGulin committed:
minor test nit

KSGulin committed:
remove config tests

KSGulin committed:
remove config tests

corey-nm reviewed (dismissed):
lgtm! 🎉

KSGulin committed:
Drop support for torch<1.9 (#138)

KSGulin committed:
Drop support for torch<1.9 (#138)

KSGulin committed:
Update directory saving flow (#139)

KSGulin committed:
Update directory saving flow (#139)

corey-nm reviewed (changes requested):
Would be good to add some tests for the best models logic for expected directory structure/saved models

KSGulin committed:
make server-dependent test optional

KSGulin committed:
make server-dependent test optional

KSGulin commented on src/sparsify/auto/api/main.py:
```diff
@@ -12,52 +12,82 @@
+    # 1. number of tuning trials used up
+    # 2. maximum tuning time exceeded
+    # 3. tuning early stopping condition met
+    while not (len(history) >= max_tune_trials) or (
+        time.time() - training_start_time > max_train_seconds
+    ):
```
Good catch

KSGulin committed:
Update src/sparsify/auto/api/main.py

Co-authored-by: corey-nm <109536191+corey-nm@users.noreply.github.com>

KSGulin committed:
Update src/sparsify/auto/api/main.py

Co-authored-by: corey-nm <109536191+corey-nm@users.noreply.github.com>

KSGulin committed:
Update src/sparsify/auto/utils/helpers.py

Co-authored-by: corey-nm <109536191+corey-nm@users.noreply.github.com>

KSGulin committed:
Test for main path + found fixes

KSGulin committed:
Test for main path + found fixes

KSGulin committed:
Merge branch 'remove_config_creator' of github.com:neuralmagic/sparsify into remove_config_creator

KSGulin committed:
minor fixes

KSGulin committed:
minor fixes

KSGulin commented on src/sparsify/auto/tasks/object_detection/yolov5/runner.py:
```diff
@@ -64,12 +63,15 @@ class Yolov5Runner(TaskRunner):
-            "checkpoint-one-shot" if self.train_args.one_shot else "last"
+            "weights/checkpoint-one-shot.pt"
+            if self.train_args.one_shot
+            else "weights/last.pt"
```
Sure, updated

KSGulin commented on src/sparsify/auto/tasks/runner.py:
```diff
@@ -139,9 +132,18 @@ class TaskRunner:
     """
 
+    # name of the field for the export model path. e.g. "model_path" for transformers
+    export_model_kwarg: Optional[str] = None
```
I agree with this in principle, but devs rather than users would be setting this value when adding a new integration. Opting to keep as is for now and revisit once alpha version is in a more stable state

KSGulin commented on src/sparsify/auto/api/main.py:
```diff
@@ -12,52 +12,82 @@
+            if len(best_n_models) == maximum_trial_saves:
+                metrics_list = [best_metrics for _, best_metrics in best_n_models]
+                remove_trial_directory(
+                    api_args.save_directory, metrics_list.index(max(metrics_list))
```
Flow has been updated and tests added to cover it. Resolving this conversation and the ones following it. Feel free to leave comments on the new flow 🙂

KSGulin committed:
1-line fix for gha test passing

KSGulin committed:
1-line fix for gha test passing

KSGulin committed:
Temporarily pinning torch to 1.9.1 for compliance with sparseml

KSGulin committed:
Temporarily pinning torch to 1.9.1 for compliance with sparseml

KSGulin commented:
@corey-nm great call on the trial management testing. Added testing of the main flow in `test_main.py`

rahul-tuli commented on tests/sparsify/auto/autosparse/test_main.py:
```diff
@@ -0,0 +1,133 @@
+    )
+    os.makedirs(deployment_path)  # create deployment dir
+    # add dummy onnx file
+    with open(os.path.join(deployment_path, "export_output.onnx"), mode="a"):
+        pass
```
Same suggestion as above

rahul-tuli commented on tests/sparsify/auto/autosparse/test_main.py:
```diff
@@ -0,0 +1,133 @@
+        pass
+
+
+@pytest.fixture(autouse=True)
+def _cleanup_directory():
+    yield
+    if os.path.exists(_SAVE_DIRECTORY):
+        shutil.rmtree(_SAVE_DIRECTORY)
+
```
nice

rahul-tuli reviewed (commented):
Minors nitpicks. Great Job!

bfineran commented on src/sparsify/auto/api/main.py:
```diff
@@ -12,52 +12,85 @@
+            runner.move_output(trial_idx)
+            best_n_trial_metrics[trial_idx] = metrics
+
+        if len(best_n_trial_metrics) > maximum_trial_saves:
```
nit: add comment or move to helper function

bfineran commented on src/sparsify/utils/constants.py:
```diff
@@ -23,13 +23,16 @@
     "get_task_info",
 ]
 
+NEURAL_MAGIC_API_ADDRESS = "http://0.0.0.0:8000"
```
yup, needs to match what we did for package

bfineran reviewed (commented):
lgtm pending comments, great work

KSGulin committed:
Update src/sparsify/auto/api/api_models.py

Co-authored-by: Rahul Tuli <rahul@neuralmagic.com>

KSGulin committed:
Address review comments

KSGulin committed:
Merge branch 'remove_config_creator' of github.com:neuralmagic/sparsify into remove_config_creator

KSGulin committed:
Centralize NM API call + nits

KSGulin commented on src/sparsify/auto/tasks/image_classification/runner.py:
```diff
@@ -217,20 +216,12 @@ def _get_metrics(self) -> Metrics:
-                ),
-            ]
-        ]
+        return os.path.join(self.train_args.save_dir, self.train_args.model_tag)
```
Depends on what is specified by the user. It is meant to conform to the IC API, which accepts relative or absolute paths. The default path is relative

KSGulin commented:
Thanks for the thorough comments. Addressed 🙂 
@rahul-tuli I updated the NM API calls to be centralized, which touched sparsify.package. Please take a look

corey-nm commented on setup.py:
```diff
@@ -31,11 +31,7 @@
-    "pyyaml>=5.0.0",
-    "click~=8.0.0",
-]
+_deps = ["pydantic>=1.8.2", "pyyaml>=5.0.0", "click~=8.0.0", "torch==1.9.1"]
```
should we do torch>=1.9.0?

corey-nm replied:
But sparseml isn't a dependency of this package right? Is pip not smart enough  when you run `pip install sparseml sparsify` to figure out that it needs 1.9.1?

KSGulin replied:
sparseml is a requirement, under `_nm_deps`

corey-nm replied:
Gotcha okay

corey-nm commented on src/sparsify/auto/api/api_models.py:
```diff
@@ -74,15 +87,31 @@ class APIArgs(BaseModel):
-            "Maximum number of hours to train before returning best trained "
-            "model. At least one model will be trained in a given run regardless of "
-            "the maximum train time set here."
+            "Maximum number of hours to train before returning best trained " "model."
```
optional nit
```suggestion
            "Maximum number of hours to train before returning best trained model."
```

corey-nm commented on src/sparsify/auto/api/api_models.py:
```diff
@@ -74,15 +87,31 @@ class APIArgs(BaseModel):
     )
+    num_trials: Optional[int] = Field(
+        title="num_trials",
+        description=(
+            "Number of tuning trials to be run before returning best found "
+            "model. max_train_time may limit the actual num_trials ran"
+        ),
```
optional nit: what does None mean for this value?

corey-nm commented on src/sparsify/auto/api/main.py:
```diff
@@ -12,52 +12,86 @@
+    # 1. number of tuning trials used up
+    # 2. maximum tuning time exceeded
+    # 3. tuning early stopping condition met
+    while (
```
flow for this is super clean now, nice! 🚀

corey-nm commented on tests/sparsify/auto/autosparse/test_main.py:
```diff
@@ -0,0 +1,132 @@
+
+# Run definition
+
+_SAVE_DIRECTORY = "./test_main"
```
Can this be done with `tmp_path` fixture from pytest?

KSGulin replied:
I do like the elegance of tmp_path, but in this case it would complicate the design, since tmp_path is only accessible within the test and the save directory is needed prior to test start

KSGulin replied:
The scope still wouldn't set it on the global namespace right? The values that are getting patched depend on save_directory, so we'd need the path defined before we patch

KSGulin replied:
Yeah I would say over-engineering at this point. But I would like to clean up this test after alpha launch 🙂

corey-nm commented on src/sparsify/auto/api/main.py:
```diff
@@ -12,52 +12,86 @@
 
 def main():
     # Parse CLI args
     api_args = APIArgs.from_cli()
```
I think this function should be split to make testing easier:
```python
def main():
    api_args = ApiArgs.from_cli()
    run(api_args)

def run(api_args: APIArgs):
    ...
```

KSGulin replied:
Hmm I see the merit in the suggestion. I'm not sure I entirely agree. I would say let's table this consideration until later. The granularity here is too fine to consider for an alpha version which will go through more product-driven revisions

corey-nm commented on tests/sparsify/auto/autosparse/test_main.py:
```diff
@@ -0,0 +1,132 @@
+        range(len(_ACCURACY_LIST)), key=lambda i: _ACCURACY_LIST[i]
+    )[-_MAXIMUM_SAVES:]
+
+    assert os.path.exists(_SAVE_DIRECTORY)
```
Can you add tests for what the save directory should contain? Will it just be the artifact directories? Will there be more?

KSGulin replied:
Good call, updated 🙂

KSGulin committed:
Update src/sparsify/auto/api/api_models.py

Co-authored-by: corey-nm <109536191+corey-nm@users.noreply.github.com>

KSGulin commented on tests/sparsify/auto/autosparse/test_main.py:
```diff
@@ -0,0 +1,132 @@
+    f"{_RUNNER_MOCK_PATH}._get_metrics",
+    MagicMock(side_effect=_METRICS_LIST),
+)
+@patch(
+    "sparsify.auto.api.main.APIArgs.from_cli",
+    MagicMock(return_value=APIArgs(**_TEST_CONFIG)),
+)
```
Same as above

KSGulin commented on src/sparsify/auto/tasks/transformers/runner.py:
```diff
@@ -52,9 +51,11 @@ class _TransformersRunner(TaskRunner):
 
     def __init__(self, config: SparsificationTrainingConfig):
         super().__init__(config)
+        self._model_save_name = ""
```
Yes, the transformers integration targets whole directories rather than specific files

KSGulin commented on src/sparsify/auto/api/api_models.py:
```diff
@@ -114,6 +143,75 @@ def from_cli(cls, args: Optional[List[str]] = None):
+    """
+
+    task: str = Field(
+        title="task",
```
Holdover from a previous architecture and no longer necessary for SparsificationTrainingConfig. Good catch!

KSGulin committed:
addess review comments

KSGulin committed:
Merge branch 'remove_config_creator' of github.com:neuralmagic/sparsify into remove_config_creator

corey-nm commented on tests/sparsify/auto/autosparse/test_main.py:
```diff
@@ -0,0 +1,140 @@
+    output_path = os.path.join(
+        _SAVE_DIRECTORY, SAVE_DIR.format(task=_TEST_CONFIG["task"])
+    )
+    assert all(
+        os.path.exists(os.path.join(output_path, folder))
+        for folder in ["run_artifacts", "logs", "deployment"]
+    )
```
Nice! Assuming there should only be MAXIMUM_SAVES sub dirs in run_artifacts, which is what the loop below is testing? Can you add a check that there isn't anything else in that dir?

Any then what's in the deployment directory again?

KSGulin replied:
Updated. The deployment directory contains whatever you need to deploy. It will always contain an .onnx file and may or may not contain sample in/out. Anything that sparseml export commands create would be contained here

corey-nm commented on tests/sparsify/auto/autosparse/test_main.py:
```diff
@@ -0,0 +1,132 @@
+def _test_trial_artifact_directory(trial_idx: int) -> bool:
+    directory_path = _TRIAL_PATH.format(trial_idx=trial_idx)
+    assert os.path.exists(directory_path), f"Directory not found: {directory_path}"
+    assert os.listdir(directory_path), f"Directory empty:  {directory_path}"
```
This is being addressed below, resolving

KSGulin committed:
add test for num trials saved

corey-nm commented on src/sparsify/auto/api/api_models.py:
```diff
@@ -74,15 +87,32 @@ class APIArgs(BaseModel):
         default=12.0,
     )
+    maximum_trial_saves: Optional[int] = Field(
+        title="maximum_trial_saves",
```
Title here as well?

KSGulin replied:
`APIArgs` does still need the title field for automatic cli arg construction. It's more of a pedantic (heh) failsafe than a hard requirement, but still nice to have

corey-nm replied:
Oooh gotcha good to know

corey-nm commented on tests/sparsify/auto/autosparse/test_main.py:
```diff
@@ -0,0 +1,141 @@
+    output_path = os.path.join(
+        _SAVE_DIRECTORY, SAVE_DIR.format(task=_TEST_CONFIG["task"])
+    )
+    assert all(
+        os.path.exists(os.path.join(output_path, folder))
+        for folder in ["run_artifacts", "logs", "deployment"]
+    )
```
```suggestion
    assert sorted(os.listdir(output_path)) == ["deployment", "logs", "run_artifacts"]
```
?

KSGulin replied:
Great suggestion

KSGulin committed:
Update tests/sparsify/auto/autosparse/test_main.py

Co-authored-by: corey-nm <109536191+corey-nm@users.noreply.github.com>

KSGulin committed:
Use TemporaryDirectory

KSGulin committed:
Merge branch 'remove_config_creator' of github.com:neuralmagic/sparsify into remove_config_creator

corey-nm reviewed (dismissed):
🥳

KSGulin committed:
Fix server test skip condition
